Add Arabic layout

mostafaayesh opened the pull request:
This request is to add Arabic Keyboard

The layout follows closely IBM PC (https://en.wikipedia.org/wiki/Arabic_keyboard#IBM_PC_Arabic_Keyboard) 
Only alteration is the letter (ذ) usually in the number row was moved to the end of the first row. This layout is the most common from my experience.

The following are a screenshots of the current layout, suggestions are welcome.

Without number row:
![without_numrow](https://user-images.githubusercontent.com/10065580/223924250-613cd47d-16b7-408f-a250-d221004a3e4e.jpg)

With number row:
![with_numrow](https://user-images.githubusercontent.com/10065580/223924200-e3ca1df9-5ae4-4a6a-bea9-15b9f31b35fd.jpg)

Julow commented on res/xml/ar_pc.xml:
```diff
@@ -0,0 +1,45 @@
+<?xml version="1.0" encoding="utf-8"?>
+<keyboard>
+  <row>
+    <key key0="ض" key1="&#1614;" key2="`" key4="esc"/>
```
If you have `esc`, I think `tab` can be useful.

Julow reviewed (approved):
Many thanks!

> Only alteration is the letter (ذ) usually in the number row was moved to the end of the first row.

This makes the keys very small. What do you think of placing it either in the bottom-left corner of `د` (currently taken by `+`) or on the second row ?

Similarly, shift doesn't seem useful and can be removed or made smaller to gain some space.

mostafaayesh commented:
I appreciate your feedback, I have made adjustments to the layout.

I also added an alternate layout similar to Samsung keyboard. Most of us using Arabic keyboards on computers or phones very rarely use the special characters.

Adjusted layout:
![msg819505937-58988](https://user-images.githubusercontent.com/10065580/224112878-408d254f-4a59-4c7e-ae4b-94ebd95b14dc.jpg)

Alternate layout (similar to Samsung keyboard)
![msg819505937-58987](https://user-images.githubusercontent.com/10065580/224112881-a1b6a540-98d8-4316-b59e-164aa188c971.jpg)

Reference Samsung keyboard layout.
![msg819505937-58986](https://user-images.githubusercontent.com/10065580/224112880-a96de699-f013-4904-98ac-12901636f815.jpg)

mostafaayesh commented:
I realized I was missing number "1" in the first row so I added that

islam2hamy commented:
Hello ,
Thank you @mostafaayesh  for your work
For me i like this one

> With number row: ![with_numrow](https://user-images.githubusercontent.com/10065580/223924200-e3ca1df9-5ae4-4a6a-bea9-15b9f31b35fd.jpg)

Also About (ذ) replace "shift" with it

islam2hamy commented:
Also , can you please make numbers like this ١ ٢ ٣ ...

Julow commented on res/xml/method.xml:
```diff
@@ -1,5 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <input-method xmlns:android="http://schemas.android.com/apk/res/android" android:settingsActivity="juloo.keyboard2.SettingsActivity" android:supportsSwitchingToNextInputMethod="true">
+  <subtype android:label="%s" android:languageTag="ar" android:imeSubtypeLocale="ar" android:imeSubtypeMode="keyboard" android:isAsciiCapable="true" android:imeSubtypeExtraValue="default_layout=ar_pc"/>
+  <subtype android:label="%s" android:languageTag="ar" android:imeSubtypeLocale="ar" android:imeSubtypeMode="keyboard" android:isAsciiCapable="true" android:imeSubtypeExtraValue="default_layout=ar_alt"/>
```
There cannot be two entry here. The `default_layout` specifies the layout that people will see by default (until they change it in the settings).

Julow reviewed (approved):
Both layouts look good. You can try the debug version of the app here: https://github.com/Julow/Unexpected-Keyboard/actions/runs/4377386493

I'll merge soon, feel free to make more changes to the layouts.

islam2hamy commented:
> Both layouts look good. You can try the debug version of the app here: https://github.com/Julow/Unexpected-Keyboard/actions/runs/4377386493
> 
> I'll merge soon, feel free to make more changes to the layouts.

Ok , it is look good
P.s parentheses need to be corrected (ex. Next video)

https://user-images.githubusercontent.com/7989030/224122224-788b4f7b-4cef-40eb-bdd0-7c335baf9d4a.mp4

mostafaayesh commented:
> > Both layouts look good. You can try the debug version of the app here: https://github.com/Julow/Unexpected-Keyboard/actions/runs/4377386493
> > I'll merge soon, feel free to make more changes to the layouts.
> 
> Ok , it is look good P.s parentheses need to be corrected (ex. Next video)

The parentheses seem to be a Right to left (RTL) issue, I will try to look if there is a solution for it.

mostafaayesh commented:
> > > Both layouts look good. You can try the debug version of the app here: https://github.com/Julow/Unexpected-Keyboard/actions/runs/4377386493
> > > I'll merge soon, feel free to make more changes to the layouts.
> > 
> > 
> > Ok , it is look good P.s parentheses need to be corrected (ex. Next video)
> 
> The parentheses seem to be a Right to left (RTL) issue, I will try to look if there is a solution for it.

I looked into it, the first character is always left to right, if you're using a letter you wouldn't notice an issue, but if you're using a bracket or parenthesis it is the wrong direction. 

In the Hebrew layout "b{" is used but that only fixes the brackets as first character and breaks them for after the direction is changed. The solution was to implement unicodes "lrm" and "rlm" as extra keys so the user can manually switch the direction but it's not very intuitive. I personally prefer not having the "b{" because it only fixes the instance of the character at the beginning of a text field in LTR, but if other people want that I don't see an issue with changing them.

I'm currently looking into a way to implement something similar to AutoCaptilization but for RTL language to fix the direction of the first character. I need some time to get familiar with the source code, if I figure it out I will open a new Pull Request for a fix that should work for all RTL languages.

In the meantime, if the layout works well for most people, I think this can be merged.

islam2hamy commented:
١ ٢ ٣ ٤ ...
PleasE

mostafaayesh commented:
> ١ ٢ ٣ ٤ ... Please

It should have the numbers now, only caveat is if you enable the number row, you will have both types.
![msg819505937-58996](https://user-images.githubusercontent.com/10065580/224143872-e4d602b2-b128-4757-aa60-2a4fa1c47ab3.jpg)

Julow reviewed (approved):
If you need both number scripts, you can use the `Fn` modifier to transform ١..٩ into 1..9. The opposite is also possible by creating new keys `ar_1`..`ar_2` that normally type 1..9 except when transformed by the `Fn` modifier.
Modifying the number row or the numpad is more involved but I would accept the changes.

> I'm currently looking into a way to implement something similar to AutoCaptilization but for RTL language to fix the direction of the first character.

That would be a wonderful addition!

I'll merge this now, thanks for your great contribution!

islam2hamy commented:
hello @mostafaayesh 
please add (  ّ ) on ( د ) button
thank you
